File: app/src/test/java/com/kickstarter/models/ProjectTest.java
package com.kickstarter.models;

import com.kickstarter.KSRobolectricTestCase;
import com.kickstarter.mock.factories.ProjectFactory;

import org.joda.time.DateTime;

import org.junit.Test;

public class ProjectTest extends KSRobolectricTestCase {
  Project projectWithSecureUrl() {
    final String projectUrl = "https://www.kickstarter.com/projects/foo/bar";
    final Project.Urls urls = Project.Urls.builder()
      .web(Project.Urls.Web.builder().project(projectUrl).rewards(projectUrl + "/rewards").build())
      .build();

    return ProjectFactory.project().toBuilder().urls(urls).build();
  }

  @Test
  public void testSecureWebProjectUrl() {
    final String projectUrl = "http://www.kickstarter.com/projects/foo/bar";

    final Project.Urls urls = Project.Urls.builder()
      .web(Project.Urls.Web.builder().project(projectUrl).rewards(projectUrl + "/rewards").build())
      .build();

    final Project project = ProjectFactory.project().toBuilder().urls(urls).build();

    assertEquals("https://www.kickstarter.com/projects/foo/bar", project.secureWebProjectUrl());
  }

  @Test
  public void testNewPledgeUrl() {
    assertEquals("https://www.kickstarter.com/projects/foo/bar/pledge/new", projectWithSecureUrl().newPledgeUrl());
  }

  @Test
  public void testEditPledgeUrl() {
    assertEquals("https://www.kickstarter.com/projects/foo/bar/pledge/edit", projectWithSecureUrl().editPledgeUrl());
  }

  @Test
  public void testPercentageFunded() {
    assertEquals(50.0f, ProjectFactory.halfWayProject().percentageFunded());
    assertEquals(100.0f, ProjectFactory.allTheWayProject().percentageFunded());
    assertEquals(200.0f, ProjectFactory.doubledGoalProject().percentageFunded());
  }

  @Test
  public void testIsApproachingDeadline() {
    final Project projectApproachingDeadline = ProjectFactory.project().toBuilder()
      .deadline(new DateTime().plusDays(1)).build();
    final Project projectNotApproachingDeadline = ProjectFactory.project().toBuilder()
      .deadline(new DateTime().plusDays(3)).build();

    assertTrue(projectApproachingDeadline.isApproachingDeadline());
    assertFalse(projectNotApproachingDeadline.isApproachingDeadline());
  }
}

File: app/src/test/java/com/kickstarter/models/ProjectTest.kt
package com.kickstarter.models

import com.kickstarter.KSRobolectricTestCase
import com.kickstarter.mock.factories.ProjectFactory.allTheWayProject
import com.kickstarter.mock.factories.ProjectFactory.doubledGoalProject
import com.kickstarter.mock.factories.ProjectFactory.halfWayProject
import com.kickstarter.mock.factories.ProjectFactory.project
import org.joda.time.DateTime
import org.junit.Test

class ProjectTest : KSRobolectricTestCase() {

    fun projectWithSecureUrl(): Project {
        val projectUrl = "https://www.kickstarter.com/projects/foo/bar"
        val urls = Urls.builder()
            .web(Web.builder().project(projectUrl).rewards("$projectUrl/rewards").build())
            .build()
        return project().toBuilder().urls(urls).build()
    }

    @Test
    fun testSecureWebProjectUrl() {
        val projectUrl = "http://www.kickstarter.com/projects/foo/bar"
        val urls = Urls.builder()
            .web(Web.builder().project(projectUrl).rewards("$projectUrl/rewards").build())
            .build()
        val project = project().toBuilder().urls(urls).build()
        assertEquals("https://www.kickstarter.com/projects/foo/bar", project.secureWebProjectUrl())
    }

    @Test
    fun testNewPledgeUrl() {
        assertEquals(
            "https://www.kickstarter.com/projects/foo/bar/pledge/new",
            projectWithSecureUrl().newPledgeUrl()
        )
    }

    @Test
    fun testEditPledgeUrl() {
        assertEquals(
            "https://www.kickstarter.com/projects/foo/bar/pledge/edit",
            projectWithSecureUrl().editPledgeUrl()
        )
    }

    @Test
    fun testPercentageFunded() {
        assertEquals(50.0f, halfWayProject().percentageFunded())
        assertEquals(100.0f, allTheWayProject().percentageFunded())
        assertEquals(200.0f, doubledGoalProject().percentageFunded())
    }

    @Test
    fun testIsApproachingDeadline() {
        val projectApproachingDeadline = project().toBuilder()
            .deadline(DateTime().plusDays(1)).build()
        val projectNotApproachingDeadline = project().toBuilder()
            .deadline(DateTime().plusDays(3)).build()
        assertTrue(projectApproachingDeadline.isApproachingDeadline)
        assertFalse(projectNotApproachingDeadline.isApproachingDeadline)
    }

    fun testEquals_whenProjectsEquals_returnTrue() {
        val projectA = project().toBuilder().id(1).build()
        val projectB = project().toBuilder().id(1).build()

        assertTrue(projectA == projectB)
    }
}
